### DISCUSSION TITLE
Tiene que existir un único método para realizar la colaboración?

### DISCUSSION THREAD

**Discussion by agmblanco** [dds-jv/foro#417] [TPA - 2024] [answered]
Buenas, cómo están? Tenemos un dilema al momento de definir la interfaz de nuestros colaboradores. No terminamos de entender si debe existir un único método `colaborar` que de alguna forma instancie la colaboración correspondiente o si podemos tener diferentes métodos para las diferentes colaboraciones, por ejemplo `colaborarConDinero`, `colaborarConVianda`, `colaborarAdministrandoHeladera`, etc. Con la precaución de que cada método debería lanzar una excepción en caso de no ser una colaboración válida para el colaborador referenciado. 
Podrían darnos una idea de por donde deberíamos ir? Gracias!

**Comment by flbulgarelli** [comment]
👋 ¡Hola!

Todas las opciones que planteás son potencialmente correctas. Elijan una opción en base a las cualidades que estuvimos trabajando, justifíquenla adecuadamente y discútanla con su ayudante el día de la corrección.

¡Saludos!

**Reply by flbulgarelli** [reply, marked as answer]
PD: la respuesta directa a su pregunta de si tiene que haber un único método es, en sentido estricto, _no_, dado que existen tres formas diferentes de colaboración. Luego, si la pregunta apunta a cómo implementar esto en objetos, como mencioné antes, no es algo que los requerimientos puedan especificar, dado que es meramente una cuestión de diseño; de ahí que les plantee que todas las opciones son _potencialmente correctas_. 

Como criterio adicional a las cualidades de diseño (piensen en términos de calidad de abstracciones, extensibilidad, simplicidad, etc), cito la heurística de [cero-uno-infinito](https://en.wikipedia.org/wiki/Zero_one_infinity_rule) que plantea (en diversos ámbitos) que en general, las relaciones están limitadas a cero, una o N instancias, y no por ejemplo a otras cantidades arbitrarias como 4, 10 o 85. 

Podríamos tratar esta heurística de forma más amplia y extenderla a la enumeración de conceptos que aparecen en un sistema: puede que haya una forma de hacer algo, puede que haya dos formas de hacer algo, pero si hay tres o más, es bastante posible que en el futuro haya aún más.

**Reply by agmblanco** [reply]
Perfecto, muchísimas gracias!
